crates/aiken-project/src/module.rs

```diff
@@ -80,12 +80,42 @@ impl ParsedModules {
             }
         }
 
+        let mut messed_up_indices = false;
+
+        // Prune the dependency graph to only keep nodes that have a path to one of our (i.e. the
+        // current project) module. This effectively prunes dependencies that are unused from the
+        // graph to ensure that we only compile the modules we actually depend on.
+        graph.retain_nodes(|graph, ix| {
+            // When discarding a node, indices in the graph end up being rewritten. Yet, we need to
+            // know starting indices for our search, so when we remove a dependency, we need find
+            // back what those indices are.
+            if messed_up_indices {
+                our_indices = HashSet::with_capacity(our_modules.len());
+                for j in graph.node_indices() {
+                    if our_modules.contains(graph[j].as_str()) {
+                        our_indices.insert(j);
+                    }
+                }
+            }
+
+            for start in our_indices.iter() {
+                if algo::astar(&*graph, *start, |end| end == ix, |_| 1, |_| 0).is_some() {
+                    messed_up_indices = false;
+                    return true;
+                }
+            }
+
+            messed_up_indices = true;
+            false
+        });
+
```

Comment:
@rvcas this is the core of the change. I ended up simply building on top of the dependency graph that we already have, to prune nodes that are deemed irrelevant. A node of the graph (i.e. module) is considered irrelevant if there's no path from any of our project modules to that other node.

Reply:
wonderful